### PR TITLE
properly validate enum class names and values

### LINKED ISSUE #7205: entity enum type validation is permissive (class name and members name)
##### **Overview of the issue**

entity generator (or import-jdl generator) will accept java reserved keywords such as `true` and `false` with the fields of java enum type. Furthermore, the class name accepts values containing invalid characters such as the comma `,`.

##### **Motivation for or Use Case**

The application will fail to start, it's tedious to recover if you didn't commit right before doing it and if you're using the entity generator, you must restart its creation.

##### **Reproduce the error**

Create an entity with reserved keyword such as `true` (through entity generator or JDL).

##### **Related issues**

Searched with the keywords `enum` and `validation` and didn't find any pertinent ones.

###### Entity generator :
1. Reference an existing method that validates java class names from the inquirer prompt that ask the enum name.
2. Implement a method that split around the character comma `,` and send resulting `String`s through an existing validation method that forbids java reserved keywords
3. Reference this new method from the `validate` field of the inquirer prompt that ask the question about the values of the enumeration.

###### JDL :
I don't know the JDL code as much as the entity generator but something similar could be done I guess. 
I know there is already something in place but if it can help, I implemented [something for a JHipster module](https://github.com/bastienmichaux/db-importer/blob/dev/lib/validation.js) that use a **single implementation** to validation inquirer prompts and a configuration file (could be the JDL if adapted).

##### **JHipster Version(s)**

4.14.0

##### **JHipster configuration**

Irrelevant

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

```json
{
    "fluentMethods": true,
    "relationships": [],
    "fields": [
        {
            "fieldName": "myenumfield",
            "fieldType": "True,false,hello",
            "fieldValues": "true,false,hello"
        }
    ],
    "changelogDate": "20180301134309",
    "dto": "no",
    "service": "no",
    "entityTableName": "boop",
    "jpaMetamodelFiltering": false,
    "pagination": "no"
}
```

##### **Browsers and Operating System**
Irrelevant

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by ruddell**
Fix #7205

- Please make sure the below checklist is followed for Pull Requests.

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/entity/prompts.js [hunk 1]
```diff
@@ -830,6 +837,7 @@ function askForField(done) {
         if (props.fieldAdd) {
             if (props.fieldIsEnum) {
                 props.fieldType = _.upperFirst(props.fieldType);
+                props.fieldValues = props.fieldValues.toUpperCase();
```
Comment (ruddell) [inline generators/entity/prompts.js]: This also fixes an issue when you run `export-jdl` but where the fieldValues contain lower-case letters.  JDL exects enum values in all caps.

File: generators/entity/prompts.js [hunk 2]
```diff
@@ -534,6 +534,9 @@ function askForField(done) {
                 } else if (jhiCore.isReservedKeyword(input, 'JAVA')) {
                     return 'Your enum name cannot contain a Java reserved keyword';
                 }
+                if (!/^[A-Za-z0-9_\s]*$/.test(input)) {
```
Comment (MathieuAA) [inline generators/entity/prompts.js]: Why whitespaces?

Comment (ruddell) [reply generators/entity/prompts.js]: Copied it from below, thanks for catching that

**Comment by MathieuAA** [comment]
@ruddell Your fix is good. However, I have two issues with the file:
  - There's a comma [here](https://github.com/jhipster/generator-jhipster/pull/7207/files#diff-59e22ebc06ff8316b80d94a3dcb8f476R561)
  - The code is duplicated.

Can I have a look before we merge it?

**Comment by MathieuAA** [comment]
Oh I get it. It's an ugly trick but with the inquirer lib from yeoman I don't think there was any choice.
Here's the workflow:
  - The user gets to enter an enum's values like
    - ENGLISH
    - ENLIGH,FRENCH
    - ENGLISH, FRENCH

Then, the input (the whole string) is made exempt of whitespaces of any kind, then split by the comma character.
My bad, your PR was perfect. I'm gonna revert the change and add a comment.

**Comment by ruddell** [comment]
I don't think we should allow spaces in the enum values unless we trim whitespace.  For example, `ENGLISH , FRENCH` breaks the Java build

**Comment by deepu105** [comment]
@ruddell we could change the regex so that space after comma is allowed but not before comma

**Comment by deepu105** [comment]
or we could just trim whitespace when splitting the values

**Comment by deepu105** [comment]
I think we should just do the later to be safe

**Comment by ruddell** [comment]
I edited the failing example, the actual issue is when the enum value itself contains spaces

**Comment by deepu105** [comment]
Then we need to change the regex

**Comment by MathieuAA** [comment]
@ruddell I don't think it comes to that, your example doesn't seem possible.

There's another solution: we change the prompt to make the user enter one enum value at a time, then clean each value.

**Comment by deepu105** [comment]
If its not possible to do in current setup then lets not worry about that

**Comment by ruddell** [comment]
@MathieuAA It is possible in the prompts since you added `\s` back.  I'm not familiar enough with regex to prevent spaces in words but allow them after commas.
screen shot 2018-03-02 at 3 23 54 pm

**Comment by MathieuAA** [comment]
@ruddell You're right!! After I entered the enum values in the prompt, the generator kept it. That's crazy

**Pull request merged**